### DISCUSSION TITLE
Use Cases for Ham2mon

### DISCUSSION THREAD

**Discussion by lordmorgul** [lordmorgul/ham2mon#8] [Ideas]
**Uses cases for ham2mon**

**Case 1 - madengr, original intent:**
**GNU Radio based SDR scanner for narrow band FM demodulation**

[https://github.com/madengr/ham2mon] (madengr ham2mon)

Narrow FM channelizing scanner for all receive bandwidth activity to monitor in terminal interface, audio output, and record to file

Example:
- typical behavior of broadband scanner device of all active channels without priority

**Case 2 - john-:**
**Embedded scanner to file utility tool**
[http://github.com/john-/ham2mon](john- ham2mon)
My primary use case is using ham2mon as an "embedded" application that requires only intermittent access via the curses gui. The user primarily interacts with the the recorded audio through a separate [https://gitlab.com/john---/cart_dashboard] (Web interface)

**Case 3 - miweber67:**
**Priority scanner with active channel audio monitoring**
[http://github.com/miweber67/ham2mon](miweber67 ham2mon)
My use case is this: Sometimes I want to hear something, anything, but I want to hear certain things over other things if they're actually happening. That means I have one demodulator and I'm feeding it to my speakers. My local repeaters are the most interesting to me, and of course I prefer some over others. If all of the repeaters are idle, perhaps I want to hear what's on simplex frequencies, but only so long as the repeaters are idle... if a channel more important to me becomes active, I want a lesser channel to be dropped in favor of the channel with "more" priority.

That implies two things to me: 1, that the frequencies in a priority file have an internal ranking, be it implicit in their order in the file or made explicit through the addition of a priority field; and 2, that the handling of priority channels needs to be reworked to avoid the apparent bug I mentioned at the top of this post.

Example:
- listen to one channel at a time (single demod) to hear preferred repeaters if active while scanning for any other traffic if priority is not active

**Case 4 - lordmorgul:**
**Fixed frequencies activity recording**
Set demodulators count and priority channels count to capture all activity on specific frequencies, such as all input/output of a repeater without missing anything.  No audio outout to speakers, silent mode with visual only indicator of current activity. Save to files and cleanup header only trash.  Separate files for not too large based on delay since new activity. Tone filtering for specific intended activity on repeaters (GMRS case in particular with channel use overlap). Prevent channelization from being near-to but not on priority channel (minor freq drift creating new file rather than append).  Creating all available frequencies to be monitored into priority channels and do not channelize other than at priority channels.

Example:

- configure a demodulator to my GMRS repeater input, set priority channel
- add additional demodulators when machine CPU can handle more
- always capture input transmissions to the repeater in the same file up to size limit

**Comment by lordmorgul** [comment]
Ideas for aligning use cases
- Case 2: curses gui updates could be paused/disabled if not needed
- Case 3: CTCSS tone filtering would need to be frequency/channel specific rather than whole IBW 
- Case 4: new dynamic channel prioritization could allow channel center to move, need to verify
- Case 3: high priority (like a "close call" scanner feature) single channel audio exclusivity (to mark a priority channel for audio exclusion, so that only specific active priority channels have volume if any priority channel is active above threshold, but if no priority channel is active then any channel could have volume.
- Case 4: Feature SSB demodulator for automated HF voice recording

**Comment by john-** [comment]
Thanks for putting this together.

Regarding Case 2:  I did mention I had a performance focus initially.  However, since I started using python3 version of ham2mon I can run a few demodulators  comfortably on an Odroid N2.  I am not concerned over any overhead from the curses interface even though for me it is unused 99% of the time.  Unless someone else has a concern over this I say impact/overhead of curses gui is not a concern.

I use ham2mon primarily for mobile [radio services](https://www.fcc.gov/wireless/bureau-divisions/mobility-division/private-land-mobile-radio-services).  These are what are used at auto racing events for car/pit and facility (corner worker) communications.  Most of the frequencies are assigned over a particular weekend.

I had to extend ham2mon with a few features that are mostly in this fork.

There is additional functionality I need that I currently address outside of ham2mon but plan on enhancing ham2mon as I feel there may be some broader interest:

1.  Classifying audio using a machine learning model to distinguish between voice, data and garbled/open microphone transmissions
-  Some [details](https://github.com/john-/ham2mon/tree/classify_v2#audio-classification) are in my branch that contains this enhancement.
3. Stepping through swaths of spectrum greater than the bandwidth my SDR supports.
- Example:  Configure ham2mon to step through 10Mhz of a band
- I have not begun this work and or thought through it

**Comment by lordmorgul** [comment]
I am interested in the classification feature combined.
The wide scan schedule would also be interesting but I will admit I do not have a driving need or use case.  But, scheduled scanning is a common way to catch overall activity in a frequency band.  Often this is used with a secondary scheduling function to extend or continue the scan when something is detected and that would align well to the classification feature.

**Reply by john-** [reply]
I am trying to get my head around the scheduled scanning feature.  This would be "at a certain day/time scan at a particular RF Freq for a specific amount of time"?

When you say "extend or continue the scan" does that mean "if there is continued activity ignore the scheduled amount of time"?

The classification feature is in my master branch.  This also includes pretty much all the changes from your master branch.  The only functional difference is I changed the implementation of "-k" (max demod length) and "-f" (min file size) options.  They are now "--min_recording" and "--max_recording" and specified in seconds.

The classification feature requires the model that can be downloaded as part of a [release](https://github.com/john-/ham2mon/releases/tag/v0.1-alpha).  Only use the `model_1.tflite` file from this as code is out of date.

**Reply by lordmorgul** [reply]
I really meant scheduled in the sense of "scan this bandwidth for this time".  For a wide scan you may want to have ability to set a scan width and time, so you could for instance spend longer looking for active signal on amateur 2m than on amateur 70cm, but regularly check both.  So a scan schedule could put 10sec in 2m band for every 3sec in 70cm band and perhaps 2sec in FRS/GMRS then back to 2m, making it check every band every 15sec.
Example: for a little while I actually ran three separate copies of ham2mon continuously to do exactly this monitoring of 2m, 70cm, GMRS while the RTL-SDRs used could not cover more than one at a time.  It also took three dongles and antennas (a splitter would have worked).  One scanner changing frequencies probably would have captured most of the same traffic because it was very sparse at my AO.

The simpler approach of "scan contiguous chunks of this really wide bandwidth equal time slots until you finish the whole bandwidth" is another viable superscan.  This is probably the more desirable overall but there are good uses for both.

Scan extension would be something like the scan delay time most radios allow setting, to stay on any active channel a fixed delta time (1-3sec is common) and then move.  The superscan will need to know if it always moves on after the prescribed delta time, or extends that time if there is activity.  For a recording channel you might want to scanner to usually move on after 1sec, but if activity then hold for 5sec and then move on.  The faster the scanner is the more critical this adjustment is, and with various SDRs there will be a big difference in how fast it retunes.  It could be that the program's main loop is the limiter but I doubt it.

**Reply by lordmorgul** [reply]
Classification would be a great decision maker on whether to keep moving too, if only set to record voice then data blurbs do not count as activity to keep the scanner range set longer.

**Reply by lordmorgul** [reply]
Scheduled Scan being as you described above, time of day for certain time, I am not really sure that making ham2mon do that is not better than just having a cronjob start ham2mon with a max run time?  Have to think thru possible use case for that.  A max application run time switch and timer before terminating the main loop, combined with "cron" or "at" would do it.

**Reply by john-** [reply]
I put together a crude flow chart.  It seems like your approach would extend this to include range (e.g. 2m) specific hang times.  In this context I mean "hang time" to be how long scanner stays at the current RF Freq without anything of interest being demodulated.

![flow_v1](https://github.com/lordmorgul/ham2mon/assets/2359125/4dde592f-00d4-428d-bf12-16ba42de55a5)

Note: I am currently doing a single range version of this outside of ham2mon.  I have an external program that sends ham2mon the keystrokes to set the RF Freq.  It does this based on whether recordings are being made within defined hang times.

**Reply by lordmorgul** [reply]
Flow chart makes sense.  I just realized when mentioning wbfm demod type that a complication of doing this scan across bands would be need to keep the same demodulator type.  Using nbfm makes sense for that use case, while some bands would be more ideal to pick wbfm, switching is too hard when the demod type is picked early in program constructor code.

The wbfm demodulator was targetted at getting full power of the wider signal in a wider channel.  It was also supposed to help pick a more accurate center frequency of the signal because narrow could dance back and forth between two channels and save half the discussion on a true frequency to two separate files because the demod frequency was not stable/repeatable for the signal.

**Reply by john-** [reply]
>  I just realized when mentioning wbfm demod type that a complication of doing this scan across bands would be need to keep the same demodulator type.

Maybe I am not understanding the terminology.  You mention wbfm in this comment.  I do not see where you mention wbfm elsewhere in this discussion.

Anyway, i agree that it would be too involved to have scan range(s) that include more than one demodulation type.  They should be only nbfm or only am.

**Reply by lordmorgul** [reply]
Sorry it is me mixing topic across discussions.  I put a wbfm issue on your repo, and mentioned here on mine.  I implemented a wbfm tuner in a branch and will bring it forward to a branch on yours after I do the priority channel logic fixes; But agree, even if adding the wbfm type as an option a scan should use all one type.

**Comment by lordmorgul** [comment]
@john-  Channel priority assignment in my fork master seems to have been brought into yours directly and is working.  But it was rudimentary. Main issue is that if too many priority channels exist for demodulators then not all priority channels can be heard (fixed assignment of priority to demod occurs).    This works fine for my use case 4 repeaters.  miweber67 improved this a year ago for his use case 3 audio monitoring , and I have not captured his update yet.  in my master. I plan to make that update to my master by merging most of his edits, but maybe in two branches, that can be similarly merged easily your way hopefully.

**Reply by lordmorgul** [reply]
As I do this merge I plan to look at whether both behaviors can be included with a switch.  For use case 4 it could be desirable to NOT record anything except a defined priority channel.
